File: regions/system/library/typescript/kafka/__tests__/kafka.test.ts
import { describe, it, expect } from 'vitest';
import {
  KafkaConfig,
  bootstrapServersString,
  usesTLS,
  validateKafkaConfig,
  DEFAULT_CONNECTION_TIMEOUT_MS,
  DEFAULT_REQUEST_TIMEOUT_MS,
  DEFAULT_MAX_MESSAGE_BYTES,
  TopicConfig,
  validateTopicName,
  topicTier,
  defaultPartitionsForTier,
  topicDefaultPartitions,
  NoOpKafkaHealthChecker,
  KafkaHealthStatus,
  KafkaError,
} from '../src/index.js';

describe('bootstrapServersString', () => {
  it('joins multiple brokers with comma', () => {
    const config: KafkaConfig = { bootstrapServers: ['broker1:9092', 'broker2:9092'], consumerGroup: 'test-group' };
    expect(bootstrapServersString(config)).toBe('broker1:9092,broker2:9092');
  });

  it('returns single broker as-is', () => {
    const config: KafkaConfig = { bootstrapServers: ['broker1:9092'], consumerGroup: 'test-group' };
    expect(bootstrapServersString(config)).toBe('broker1:9092');
  });
});

describe('usesTLS', () => {
  it('returns true for SSL', () => {
    const config: KafkaConfig = { bootstrapServers: ['b:9092'], securityProtocol: 'SSL', consumerGroup: 'test-group' };
    expect(usesTLS(config)).toBe(true);
  });

  it('returns true for SASL_SSL', () => {
    const config: KafkaConfig = { bootstrapServers: ['b:9092'], securityProtocol: 'SASL_SSL', consumerGroup: 'test-group' };
    expect(usesTLS(config)).toBe(true);
  });

  it('returns false for PLAINTEXT', () => {
    const config: KafkaConfig = { bootstrapServers: ['b:9092'], securityProtocol: 'PLAINTEXT', consumerGroup: 'test-group' };
    expect(usesTLS(config)).toBe(false);
  });

  it('returns false for SASL_PLAINTEXT', () => {
    const config: KafkaConfig = { bootstrapServers: ['b:9092'], securityProtocol: 'SASL_PLAINTEXT', consumerGroup: 'test-group' };
    expect(usesTLS(config)).toBe(false);
  });
});

describe('validateKafkaConfig', () => {
  it('passes with valid config', () => {
    const config: KafkaConfig = { bootstrapServers: ['broker1:9092'], securityProtocol: 'PLAINTEXT', consumerGroup: 'test-group' };
    expect(() => validateKafkaConfig(config)).not.toThrow();
  });

  it('throws on empty bootstrapServers', () => {
    const config: KafkaConfig = { bootstrapServers: [], consumerGroup: 'test-group' };
    expect(() => validateKafkaConfig(config)).toThrow('bootstrap servers must not be empty');
  });

  it('throws on invalid protocol', () => {
    const config = { bootstrapServers: ['broker1:9092'], securityProtocol: 'INVALID', consumerGroup: 'test-group' } as unknown as KafkaConfig;
    expect(() => validateKafkaConfig(config)).toThrow('invalid security protocol');
  });

  it('throws on empty consumerGroup', () => {
    const config: KafkaConfig = { bootstrapServers: ['broker1:9092'], consumerGroup: '' };
    expect(() => validateKafkaConfig(config)).toThrow('consumer group must not be empty');
  });

  it('throws on non-positive connectionTimeoutMs', () => {
    const config: KafkaConfig = { bootstrapServers: ['broker1:9092'], consumerGroup: 'test-group', connectionTimeoutMs: 0 };
    expect(() => validateKafkaConfig(config)).toThrow('connection timeout must be positive');
  });

  it('throws on non-positive requestTimeoutMs', () => {
    const config: KafkaConfig = { bootstrapServers: ['broker1:9092'], consumerGroup: 'test-group', requestTimeoutMs: -1 };
    expect(() => validateKafkaConfig(config)).toThrow('request timeout must be positive');
  });

  it('throws on non-positive maxMessageBytes', () => {
    const config: KafkaConfig = { bootstrapServers: ['broker1:9092'], consumerGroup: 'test-group', maxMessageBytes: 0 };
    expect(() => validateKafkaConfig(config)).toThrow('max message bytes must be positive');
  });

  it('passes with all optional timeout fields set', () => {
    const config: KafkaConfig = {
      bootstrapServers: ['broker1:9092'],
      consumerGroup: 'test-group',
      connectionTimeoutMs: 10000,
      requestTimeoutMs: 60000,
      maxMessageBytes: 2_000_000,
    };
    expect(() => validateKafkaConfig(config)).not.toThrow();
  });
});

describe('default constants', () => {
  it('has correct default connection timeout', () => {
    expect(DEFAULT_CONNECTION_TIMEOUT_MS).toBe(5000);
  });

  it('has correct default request timeout', () => {
    expect(DEFAULT_REQUEST_TIMEOUT_MS).toBe(30000);
  });

  it('has correct default max message bytes', () => {
    expect(DEFAULT_MAX_MESSAGE_BYTES).toBe(1_000_000);
  });
});

describe('validateTopicName', () => {
  const validNames = [
    'k1s0.system.user.created.v1',
    'k1s0.business.order.placed.v2',
    'k1s0.service.payment.processed.v1',
    'k1s0.system.user-profile.updated.v10',
    'k1s0.system.auth.token-refreshed.v1',
  ];

  it.each(validNames)('passes for valid topic name: %s', (name) => {
    expect(() => validateTopicName({ name })).not.toThrow();
  });

  const invalidNames = [
    '',
    'invalid',
    'k1s0.invalid.user.created.v1',
    'k1s0.system.USER.created.v1',
    'k1s0.system.user.created',
    'k1s0.system.user.created.v',
  ];

  it.each(invalidNames)('throws for invalid topic name: "%s"', (name) => {
    expect(() => validateTopicName({ name })).toThrow();
  });
});

describe('topicTier', () => {
  it('returns "system" for system topic', () => {
    expect(topicTier({ name: 'k1s0.system.user.created.v1' })).toBe('system');
  });

  it('returns "business" for business topic', () => {
    expect(topicTier({ name: 'k1s0.business.order.placed.v1' })).toBe('business');
  });

  it('returns "service" for service topic', () => {
    expect(topicTier({ name: 'k1s0.service.payment.done.v1' })).toBe('service');
  });

  it('returns empty string for invalid topic name', () => {
    expect(topicTier({ name: 'invalid-name' })).toBe('');
  });
});

describe('defaultPartitionsForTier', () => {
  it('returns 6 for system tier', () => {
    expect(defaultPartitionsForTier('system')).toBe(6);
  });

  it('returns 6 for business tier', () => {
    expect(defaultPartitionsForTier('business')).toBe(6);
  });

  it('returns 3 for service tier', () => {
    expect(defaultPartitionsForTier('service')).toBe(3);
  });

  it('returns 3 for unknown tier', () => {
    expect(defaultPartitionsForTier('other')).toBe(3);
  });
});

describe('topicDefaultPartitions', () => {
  it('returns 6 for system topic', () => {
    expect(topicDefaultPartitions({ name: 'k1s0.system.auth.login.v1' })).toBe(6);
  });

  it('returns 6 for business topic', () => {
    expect(topicDefaultPartitions({ name: 'k1s0.business.order.placed.v1' })).toBe(6);
  });

  it('returns 3 for service topic', () => {
    expect(topicDefaultPartitions({ name: 'k1s0.service.payment.done.v1' })).toBe(3);
  });

  it('returns 3 for invalid topic name', () => {
    expect(topicDefaultPartitions({ name: 'invalid' })).toBe(3);
  });
});

describe('NoOpKafkaHealthChecker', () => {
  it('returns configured healthy status', async () => {
    const status: KafkaHealthStatus = { healthy: true, message: 'OK', brokerCount: 3 };
    const checker = new NoOpKafkaHealthChecker(status);
    const result = await checker.healthCheck();
    expect(result.healthy).toBe(true);
    expect(result.brokerCount).toBe(3);
  });

  it('returns configured unhealthy status', async () => {
    const status: KafkaHealthStatus = { healthy: false, message: 'connection refused', brokerCount: 0 };
    const checker = new NoOpKafkaHealthChecker(status);
    const result = await checker.healthCheck();
    expect(result.healthy).toBe(false);
  });

  it('throws when error is configured', async () => {
    const status: KafkaHealthStatus = { healthy: false, message: '', brokerCount: 0 };
    const err = new Error('connection refused');
    const checker = new NoOpKafkaHealthChecker(status, err);
    await expect(checker.healthCheck()).rejects.toThrow('connection refused');
  });
});

describe('KafkaError', () => {
  it('has correct name and message', () => {
    const err = new KafkaError('test error');
    expect(err.name).toBe('KafkaError');
    expect(err.message).toBe('test error');
    expect(err).toBeInstanceOf(Error);
  });
});

File: regions/system/library/dart/kafka/test/kafka_test.dart
import 'package:test/test.dart';
import 'package:k1s0_kafka/kafka.dart';

void main() {
  group('KafkaConfig', () {
    group('bootstrapServersString', () {
      test('複数のブローカーをカンマで結合すること', () {
        final config = KafkaConfig(
          bootstrapServers: ['broker1:9092', 'broker2:9092', 'broker3:9092'],
          consumerGroup: 'test-group',
        );
        expect(config.bootstrapServersString(),
            equals('broker1:9092,broker2:9092,broker3:9092'));
      });

      test('単一ブローカーをそのまま返すこと', () {
        final config = KafkaConfig(
          bootstrapServers: ['broker1:9092'],
          consumerGroup: 'test-group',
        );
        expect(config.bootstrapServersString(), equals('broker1:9092'));
      });
    });

    group('usesTLS', () {
      test('SSLの場合にtrueを返すこと', () {
        final config = KafkaConfig(
          bootstrapServers: ['broker:9092'],
          securityProtocol: 'SSL',
          consumerGroup: 'test-group',
        );
        expect(config.usesTLS(), isTrue);
      });

      test('SASL_SSLの場合にtrueを返すこと', () {
        final config = KafkaConfig(
          bootstrapServers: ['broker:9092'],
          securityProtocol: 'SASL_SSL',
          consumerGroup: 'test-group',
        );
        expect(config.usesTLS(), isTrue);
      });

      test('PLAINTEXTの場合にfalseを返すこと', () {
        final config = KafkaConfig(
          bootstrapServers: ['broker:9092'],
          securityProtocol: 'PLAINTEXT',
          consumerGroup: 'test-group',
        );
        expect(config.usesTLS(), isFalse);
      });

      test('SASL_PLAINTEXTの場合にfalseを返すこと', () {
        final config = KafkaConfig(
          bootstrapServers: ['broker:9092'],
          securityProtocol: 'SASL_PLAINTEXT',
          consumerGroup: 'test-group',
        );
        expect(config.usesTLS(), isFalse);
      });

      test('securityProtocolがnullの場合にfalseを返すこと', () {
        final config = KafkaConfig(
          bootstrapServers: ['broker:9092'],
          consumerGroup: 'test-group',
        );
        expect(config.usesTLS(), isFalse);
      });
    });

    group('validate', () {
      test('bootstrapServersが空の場合にエラーを投げること', () {
        final config = KafkaConfig(bootstrapServers: [], consumerGroup: 'test-group');
        expect(
          () => config.validate(),
          throwsA(isA<KafkaError>().having(
              (e) => e.message, 'message', contains('bootstrap servers'))),
        );
      });

      test('不正なセキュリティプロトコルでエラーを投げること', () {
        final config = KafkaConfig(
          bootstrapServers: ['broker:9092'],
          securityProtocol: 'INVALID',
          consumerGroup: 'test-group',
        );
        expect(
          () => config.validate(),
          throwsA(isA<KafkaError>().having(
              (e) => e.message, 'message', contains('invalid security'))),
        );
      });

      test('有効なPLAINTEXT設定でバリデーションが通ること', () {
        final config = KafkaConfig(
          bootstrapServers: ['broker:9092'],
          securityProtocol: 'PLAINTEXT',
          consumerGroup: 'test-group',
        );
        expect(() => config.validate(), returnsNormally);
      });

      test('有効なSSL設定でバリデーションが通ること', () {
        final config = KafkaConfig(
          bootstrapServers: ['broker:9092'],
          securityProtocol: 'SSL',
          consumerGroup: 'test-group',
        );
        expect(() => config.validate(), returnsNormally);
      });

      test('有効なSASL_PLAINTEXT設定でバリデーションが通ること', () {
        final config = KafkaConfig(
          bootstrapServers: ['broker:9092'],
          securityProtocol: 'SASL_PLAINTEXT',
          consumerGroup: 'test-group',
        );
        expect(() => config.validate(), returnsNormally);
      });

      test('有効なSASL_SSL設定でバリデーションが通ること', () {
        final config = KafkaConfig(
          bootstrapServers: ['broker:9092'],
          securityProtocol: 'SASL_SSL',
          consumerGroup: 'test-group',
        );
        expect(() => config.validate(), returnsNormally);
      });

      test('securityProtocolなしでバリデーションが通ること', () {
        final config = KafkaConfig(
          bootstrapServers: ['broker:9092'],
          consumerGroup: 'test-group',
        );
        expect(() => config.validate(), returnsNormally);
      });

      test('consumerGroupが空の場合にエラーを投げること', () {
        final config = KafkaConfig(
          bootstrapServers: ['broker:9092'],
          consumerGroup: '',
        );
        expect(
          () => config.validate(),
          throwsA(isA<KafkaError>().having(
              (e) => e.message, 'message', contains('consumer group'))),
        );
      });

      test('connectionTimeoutMsが0以下の場合にエラーを投げること', () {
        final config = KafkaConfig(
          bootstrapServers: ['broker:9092'],
          consumerGroup: 'test-group',
          connectionTimeoutMs: 0,
        );
        expect(
          () => config.validate(),
          throwsA(isA<KafkaError>().having(
              (e) => e.message, 'message', contains('connection timeout'))),
        );
      });

      test('requestTimeoutMsが0以下の場合にエラーを投げること', () {
        final config = KafkaConfig(
          bootstrapServers: ['broker:9092'],
          consumerGroup: 'test-group',
          requestTimeoutMs: -1,
        );
        expect(
          () => config.validate(),
          throwsA(isA<KafkaError>().having(
              (e) => e.message, 'message', contains('request timeout'))),
        );
      });

      test('maxMessageBytesが0以下の場合にエラーを投げること', () {
        final config = KafkaConfig(
          bootstrapServers: ['broker:9092'],
          consumerGroup: 'test-group',
          maxMessageBytes: 0,
        );
        expect(
          () => config.validate(),
          throwsA(isA<KafkaError>().having(
              (e) => e.message, 'message', contains('max message bytes'))),
        );
      });

      test('全オプションフィールド指定でバリデーションが通ること', () {
        final config = KafkaConfig(
          bootstrapServers: ['broker:9092'],
          consumerGroup: 'test-group',
          connectionTimeoutMs: 10000,
          requestTimeoutMs: 60000,
          maxMessageBytes: 2000000,
        );
        expect(() => config.validate(), returnsNormally);
      });
    });

    group('デフォルト値', () {
      test('connectionTimeoutMsのデフォルト値が5000であること', () {
        final config = KafkaConfig(
          bootstrapServers: ['broker:9092'],
          consumerGroup: 'test-group',
        );
        expect(config.connectionTimeoutMs, equals(5000));
      });

      test('requestTimeoutMsのデフォルト値が30000であること', () {
        final config = KafkaConfig(
          bootstrapServers: ['broker:9092'],
          consumerGroup: 'test-group',
        );
        expect(config.requestTimeoutMs, equals(30000));
      });

      test('maxMessageBytesのデフォルト値が1000000であること', () {
        final config = KafkaConfig(
          bootstrapServers: ['broker:9092'],
          consumerGroup: 'test-group',
        );
        expect(config.maxMessageBytes, equals(1000000));
      });
    });
  });

  group('TopicConfig', () {
    group('validateName', () {
      test('有効なsystemトピック名を受け入れること', () {
        final topic =
            TopicConfig(name: 'k1s0.system.user.created.v1');
        expect(() => topic.validateName(), returnsNormally);
      });

      test('有効なbusinessトピック名を受け入れること', () {
        final topic =
            TopicConfig(name: 'k1s0.business.order.placed.v1');
        expect(() => topic.validateName(), returnsNormally);
      });

      test('有効なserviceトピック名を受け入れること', () {
        final topic =
            TopicConfig(name: 'k1s0.service.payment.completed.v2');
        expect(() => topic.validateName(), returnsNormally);
      });

      test('ハイフンを含むトピック名を受け入れること', () {
        final topic = TopicConfig(
            name: 'k1s0.system.user-auth.session-created.v1');
        expect(() => topic.validateName(), returnsNormally);
      });

      test('数字を含むトピック名を受け入れること', () {
        final topic =
            TopicConfig(name: 'k1s0.system.user123.created456.v10');
        expect(() => topic.validateName(), returnsNormally);
      });

      test('空のトピック名でエラーを投げること', () {
        final topic = TopicConfig(name: '');
        expect(
          () => topic.validateName(),
          throwsA(isA<KafkaError>()
              .having((e) => e.message, 'message', contains('empty'))),
        );
      });

      test('プレフィックスが欠けている場合にエラーを投げること', () {
        final topic = TopicConfig(name: 'invalid.system.user.created.v1');
        expect(
          () => topic.validateName(),
          throwsA(isA<KafkaError>()
              .having((e) => e.message, 'message', contains('invalid topic'))),
        );
      });

      test('不正なティアでエラーを投げること', () {
        final topic = TopicConfig(name: 'k1s0.invalid.user.created.v1');
        expect(
          () => topic.validateName(),
          throwsA(isA<KafkaError>()),
        );
      });

      test('バージョンが欠けている場合にエラーを投げること', () {
        final topic = TopicConfig(name: 'k1s0.system.user.created');
        expect(
          () => topic.validateName(),
          throwsA(isA<KafkaError>()),
        );
      });

      test('大文字が含まれる場合にエラーを投げること', () {
        final topic = TopicConfig(name: 'k1s0.system.User.Created.v1');
        expect(
          () => topic.validateName(),
          throwsA(isA<KafkaError>()),
        );
      });

      test('セグメントが不足している場合にエラーを投げること', () {
        final topic = TopicConfig(name: 'k1s0.system.user.v1');
        expect(
          () => topic.validateName(),
          throwsA(isA<KafkaError>()),
        );
      });
    });

    group('tier', () {
      test('systemティアを返すこと', () {
        final topic =
            TopicConfig(name: 'k1s0.system.user.created.v1');
        expect(topic.tier(), equals('system'));
      });

      test('businessティアを返すこと', () {
        final topic =
            TopicConfig(name: 'k1s0.business.order.placed.v1');
        expect(topic.tier(), equals('business'));
      });

      test('serviceティアを返すこと', () {
        final topic =
            TopicConfig(name: 'k1s0.service.payment.completed.v1');
        expect(topic.tier(), equals('service'));
      });

      test('不正なトピック名で空文字列を返すこと', () {
        final topic = TopicConfig(name: 'invalid');
        expect(topic.tier(), equals(''));
      });
    });

    group('partitionsForTier', () {
      test('systemティアで6を返すこと', () {
        expect(TopicConfig.partitionsForTier('system'), equals(6));
      });

      test('businessティアで6を返すこと', () {
        expect(TopicConfig.partitionsForTier('business'), equals(6));
      });

      test('serviceティアで3を返すこと', () {
        expect(TopicConfig.partitionsForTier('service'), equals(3));
      });

      test('不明なティアで3を返すこと', () {
        expect(TopicConfig.partitionsForTier('other'), equals(3));
      });
    });

    group('defaultPartitionsForTier', () {
      test('systemトピックで6を返すこと', () {
        final topic =
            TopicConfig(name: 'k1s0.system.auth.login.v1');
        expect(topic.defaultPartitionsForTier(), equals(6));
      });

      test('businessトピックで6を返すこと', () {
        final topic =
            TopicConfig(name: 'k1s0.business.order.placed.v1');
        expect(topic.defaultPartitionsForTier(), equals(6));
      });

      test('serviceトピックで3を返すこと', () {
        final topic =
            TopicConfig(name: 'k1s0.service.payment.done.v1');
        expect(topic.defaultPartitionsForTier(), equals(3));
      });

      test('不正なトピック名で3を返すこと', () {
        final topic = TopicConfig(name: 'invalid');
        expect(topic.defaultPartitionsForTier(), equals(3));
      });
    });
  });

  group('KafkaHealthStatus', () {
    test('必須フィールドで生成できること', () {
      const status = KafkaHealthStatus(
        healthy: true,
        message: 'OK',
        brokerCount: 3,
      );
      expect(status.healthy, isTrue);
      expect(status.message, equals('OK'));
      expect(status.brokerCount, equals(3));
    });
  });

  group('NoOpKafkaHealthChecker', () {
    test('設定済みステータスを返すこと', () async {
      const status = KafkaHealthStatus(
        healthy: true,
        message: 'All brokers connected',
        brokerCount: 3,
      );
      final checker = NoOpKafkaHealthChecker(status: status);
      final result = await checker.healthCheck();
      expect(result.healthy, isTrue);
      expect(result.message, equals('All brokers connected'));
      expect(result.brokerCount, equals(3));
    });

    test('設定済みエラーを投げること', () async {
      const status = KafkaHealthStatus(
        healthy: false,
        message: 'error',
        brokerCount: 0,
      );
      final checker = NoOpKafkaHealthChecker(
        status: status,
        error: Exception('connection failed'),
      );
      expect(
        () => checker.healthCheck(),
        throwsA(isA<Exception>()),
      );
    });
  });

  group('KafkaError', () {
    test('toStringにメッセージが含まれること', () {
      const err = KafkaError('test error');
      expect(err.toString(), contains('test error'));
    });
  });
}
